rts/motoko-rts/src/stabilization/layout.rs

```diff
@@ -0,0 +1,404 @@
+//! The long-term layout definition of the stable format.
+//!
+//! The stable object graph resides a linear stable memory space.
+//!
+//! Pointers are serialized as 64-bit skewed offsets in that space.
+//! The stable encoding of scalars is currently identical to the
+//! main memory encoding, requiring the precise scalar tag as metadata.
+//! If the main memory scalar representation changes in future, the
+//! values need to be explicitly translated to the expected stable format.
+//!
+//! Each object uses a `StableTag` as header and is followed by
+//! the object payload as outlined in the corresponding Rust structs.
+//! Some objects, such as `StableArray`, `StableObject`, `StableBlob`,
+//! and `StableBigNum` have a dynamic payload body in an addition to a static
+//! header.
+//!
+//! Not all heap memory object types are stabilized because some
+//! of them are not stable types. New object types can be added
+//! with backwards compatibility but encoding changes to existing stable
+//! data types must be handled with extra care to ensure backwards compatibility.
+//!
+//! Note: For a potential downgrade to 32-bit, the 64-bit stable pointer
+//! offsets can be scaled down by a factor `8` during the destabilization
+//! such that they fit into 32-bit values during Cheney's graph-copy.
+
+use crate::{
+    barriers::allocation_barrier,
+    constants::WORD_SIZE,
+    memory::Memory,
+    rts_trap_with,
+    types::{
+        size_of, Tag, Value, TAG_ARRAY, TAG_ARRAY_SLICE_MIN, TAG_BIGINT, TAG_BITS64, TAG_BLOB,
+        TAG_CONCAT, TAG_MUTBOX, TAG_OBJECT, TAG_OBJ_IND, TAG_REGION, TAG_SOME, TAG_VARIANT,
+        TRUE_VALUE,
+    },
+};
+
+use self::{
+    stable_array::StableArray, stable_bigint::StableBigInt, stable_bits64::StableBits64,
+    stable_blob::StableBlob, stable_concat::StableConcat, stable_mutbox::StableMutBox,
+    stable_obj_ind::StableObjInd, stable_object::StableObject, stable_region::StableRegion,
+    stable_some::StableSome, stable_variant::StableVariant,
+};
+
+use super::{
+    deserialization::stable_memory_access::StableMemoryAccess,
+    serialization::{
+        stable_memory_stream::{ScanStream, StableMemoryStream, WriteStream},
+        SerializationContext,
+    },
+};
+
+mod stable_array;
+mod stable_bigint;
+mod stable_bits64;
+mod stable_blob;
+mod stable_concat;
+mod stable_mutbox;
+mod stable_obj_ind;
+mod stable_object;
+mod stable_region;
+mod stable_some;
+mod stable_variant;
+
+/// Different kinds of objects used in the stable format.
+/// Note: The tags are encoded as `StableTag` to be represented as `u64`.
+/// Native Rust enums would be encoded as `i32`, leading to unwanted alignment
+/// when directly serializing them as tags to the 64-bit stable format.
+#[repr(C)]
+#[derive(Clone, Copy, PartialEq)]
+pub enum StableObjectKind {
```

Comment:
I wonder if you could use
```
[#repr(u64)] 
pub enum ...
  Array = 1, 
```
etc to avoid declaring u64 constants `STABLE_TAG_ARRAY` but perhaps that doesn't work for u64...

Reply:
Actually, it is a good idea. I refactored it accordingly. Thanks for the suggestion. The only remaining problem is that the conversion from `u64` to enum still requires a custom, quite cumbersome, mapping function using constants. (It seems that there is no easy approach, maybe some libraries, but it is not worth taking that complexity/risk).